cpp/src/arrow/dataset/file_base.cc

```diff
@@ -41,5 +44,54 @@ Status FileBasedDataFragment::GetTasks(std::shared_ptr<ScanContext> scan_context
   return format_->ScanFile(location_, scan_options_, scan_context, out);
 }
 
+DirectoryBasedDataSource::DirectoryBasedDataSource(const FileSource& location,
+                                                   std::vector<fs::FileStats> stats,
+                                                   std::shared_ptr<FileFormat> format,
+                                                   std::shared_ptr<ScanOptions> opts)
+    : location_(location),
+      stats_(std::move(stats)),
+      format_(std::move(format)),
+      scan_options_(std::move(opts)) {}
+
+Status DirectoryBasedDataSource::Make(const FileSource& location,
+                                      std::shared_ptr<FileFormat> format,
+                                      std::shared_ptr<ScanOptions> opts,
+                                      std::unique_ptr<DirectoryBasedDataSource>* out) {
+  std::vector<fs::FileStats> stats;
+  fs::Selector s;
+  s.base_dir = location.path();
+  s.recursive = true;
+  RETURN_NOT_OK(location.filesystem()->GetTargetStats(s, &stats));
+
+  auto new_end =
+      std::remove_if(stats.begin(), stats.end(), [&](const fs::FileStats& stats) {
+        return stats.type() != fs::FileType::File ||
+               !format->IsKnownExtension(stats.extension());
+      });
+  stats.resize(new_end - stats.begin());
+
+  out->reset(new DirectoryBasedDataSource(location, std::move(stats), std::move(format),
+                                          std::move(opts)));
+  return Status::OK();
+}
+
+std::unique_ptr<DataFragmentIterator> DirectoryBasedDataSource::GetFragments(
+    std::shared_ptr<ScanOptions> options) {
+  DataFragmentVector fragments;
+
+  for (const auto& stats : stats_) {
+    FileSource l(stats.path(), location_.filesystem());
+    std::unique_ptr<FileBasedDataFragment> fragment;
+    auto st = format_->MakeFragment(l, scan_options_, &fragment);
+    if (!st.ok()) {
+      return internal::make_unique<EmptyIterator<std::shared_ptr<DataFragment>>>(st);
```

Comment:
Hum, I'm not sure I agree. Our choices are:

- Change the signature to return failure (possible)
- Construct the DataFragments at construction, the downside is that a data fragment and/or data source might only be consumed once. This is a general problem we might have to take a decision.

The solution to returns an EmptyIterator seems like a hack. I would think the first option is probably better, because the data source _can_ fail due to external reasons, e.g. file system activity.

Reply:
I don't think the second option is viable; if DataSources are single use then that would apply to Datasets as well. Would it be more acceptable to lazily make fragments in the iterator (rather than making them all ahead of time)? I can also change the return type to `Result<ptr<DataFragmentIterator>>` if necessary, but it seems reasonable to defer error handling to the data fragment consumer.

Reply:
I don't think the second option is viable; if DataSources are single use then that would apply to Datasets as well. Would it be more acceptable to lazily make fragments in the iterator (rather than making them all ahead of time)? I can also change the return type to Result<ptr<DataFragmentIterator>> if necessary, but it seems reasonable to defer error handling to the data fragment consumer.

Reply:
I think we should either use all status or all result, but not a mix.

Reply:
Also note that while it must yield an iterator, the failure could happen in the first call of Next.

Reply:
I think I prefer this version:

- Combine VectorIterator(paths) with MapIterator(pathToFragment, paths_it)

Reply:
done